docs/cvl/cvl2/changes.md

```diff
@@ -0,0 +1,583 @@
+not exist in the contract; rules that would call the non-existent method are
+skipped during verification.
+
+This behavior can lead to confusion, because typos or name changes could silently
+cause a rule to be skipped.
+
+In CVL 2, this behavior is still available, but the methods entry must contain
+the keyword `optional` somewhere after the `returns` clause and before the
+summarization (if any).
+
+```{todo}
+If a methods block contains a non-optional entry for a method that doesn't exist
+in the contract, you will receive the following error message:
+```
+
+### `library` annotations
+
+In CVL 2, contract functions declared as library functions must be annotated
+with `library` in the `methods` block.
+
+```{todo}
+If you forget to declare a method as a `library` method, you will receive the
+following error message:
+```
+
+### Required `calldata`, `memory`, or `storage` annotations for reference types
+
+In CVL 2, methods entries for internal functions must contain either `calldata`,
+`memory`, or `storage` annotations for all arguments with reference types (such
+as arrays).
+
+```{todo}
+is `calldata` actually one of the options?
+```
+
+```{todo}
+If you do not change this, you will see the following error:
+```
+
+### Summaries only apply to one contract by default
+
+In CVL 1, a summary in the `methods` block applied to all methods with the
+given signature.  Entries that had both an explicit receiver and a summary,
+such as the following, were disallowed:
+
+```cvl
+using C as c
+
+methods {
+    c.f(uint) => NONDET
+}
+```
+
+In CVL 2, summaries only apply to a single contract, unless the old behavior is
+explicitly requested by using `_` as the receiver.  If no contract is specified,
+the default is `currentContract`.
+
+Consider the following example:
+```cvl
+using C as c;
+
+methods {
+    function f(uint)   => NONDET;
+    function c.g(uint) => ALWAYS(4);
+    function h(uint)   => ALWAYS(1);
+    function _.h(uint) => NONDET;
+}
+```
+
+In this example, `currentContract.f` has a `NONDET` summary, `c.g` has an `ALWAYS`
+summary, a call to `currentContact.h` has an `ALWAYS` summary and a call to
+`h(uint)` on any other contract will use a `NONDET` summary.
+
+Summaries for specific contract methods (including the default
+`currentContract`) always override wildcard summaries.
+
+Wildcard entries cannot be declared `optional` or `envfree`, since these
+annotations only make sense for specific contract methods.
+
+```{warning}
+The meaning of your summarizations will change from CVL 1 to CVL 2.  In CVL 2,
+any entry without a `_` will only apply to a single contract!
+```
+
+### Requirements on `returns`
+
+In CVL 2, methods block entries require a `returns` clause in the following
+situations:
+
+ - The method is not a wildcard entry and the contract function returns a value.
+   In this case, the methods block entry must match the contract function's
+   return type.
```

Comment:
wait really? you can't just do `foo(uint) envfree`?

Reply:
Also, I'm not sure what this means. If the contract function returns something you need to have a return annotation. So, isn't that always?

Reply:
I don't think you need to (or can) explicitly indicate that something void?